setuptools/dist.py

```diff
@@ -212,8 +212,11 @@ def check_importable(dist, attr, value):
 
 
 def assert_string_list(dist, attr, value):
-    """Verify that value is a string list or None"""
+    """Verify that value is a string list"""
     try:
+        # verify that value is not a single-use iterable
+        assert list(value) == list(value)
```

Comment:
@benoit-pierre does this seem like the right implementation to check whether value is consumable?

For my records, [this was](https://treyhunner.com/2018/02/python-range-is-not-an-iterator/) a helpful reference here.

Reply:
I wonder if we should just simplify the whole thing and just explicitly check for a list/tuple with `isinstance(value, (tuple, list))`. @jaraco, @pganssle?

Reply:
I think using `isinstance` in this case is better than consuming the list and checking if it's consumed for sure.

I haven't had time to look at the rest of this PR and see where this `assert_string_list` is actually used - is there any chance that people would have legitimate reasons for passing a `set` to something that calls this?

Reply:
Currently, `assert_string_list` is called by `dist.check_nsp`, `dist.check_package_data` (added by this PR), and directly for some of the following:

https://github.com/pypa/setuptools/blob/3adda966e2e76ad59cf79fd8bdeab77fb11d308f/setup.py#L92-L114

https://github.com/pypa/setuptools/blob/49a839f8905e548b2cd34ba0077595e4493c5900/setuptools/dist.py#L225-L228

So I think the total list of setup keywords that end up hitting `assert_string_list` is:

+ `namespace_packages`
+ `namespace_packages`
+ `namespace_packages`
+ `namespace_packages`

Reply:
> I wonder if we should just simplify the whole thing and just explicitly check for a list/tuple with `isinstance(value, (tuple, list))`

Implemented in https://github.com/pypa/setuptools/pull/1769/commits/d9035efb4ba0e1abeb4ee25954c208b558d786d2

Reply:
Pinging @benoit-pierre and @pganssle